File: algorithms/2_medium/S/03867_sum_of_gcd_of_formed_pairs/solution.c
static inline int gcd_int(int a, int b)
{
    assert(a >= 0 && b >= 0);
    while (b != 0) {
        const int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

static inline int max_int(const int a, const int b)
{
    return a > b ? a : b;
}

static int cmp(const void* a_ptr, const void* b_ptr)
{
    const int a = *(int*)a_ptr;
    const int b = *(int*)b_ptr;
    return (a > b) - (a < b);
}

long long gcdSum(int* nums, const int len)
{
    assert(len > 0);
    int max = 0;

    for (int i = 0; i < len; i += 1) {
        max = max_int(max, nums[i]);
        nums[i] = gcd_int(max, nums[i]);
    }

    qsort(nums, (size_t)len, sizeof(*nums), cmp);

    int lo = 0;
    int hi = len - 1;
    long long sum = 0;

    while (lo < hi) {
        sum += gcd_int(nums[lo], nums[hi]);
        lo += 1;
        hi -= 1;
    }

    return sum;
}

File: algorithms/2_medium/S/03867_sum_of_gcd_of_formed_pairs/solution.rs
impl Solution {
    #[inline(always)]
    fn gcd(mut a: i32, mut b: i32) -> i32 {
        assert!(a >= 0 && b >= 0);
        while b != 0 {
            let temp = b;
            b = a % b;
            a = temp;
        }
        a
    }

    pub fn gcd_sum(mut nums: Vec<i32>) -> i64 {
        assert!(nums.len() > 0);
        let mut max = 0;

        for num in nums.iter_mut() {
            max = max.max(*num);
            *num = Self::gcd(max, *num);
        }

        nums.sort_unstable();

        let (mut lo, mut hi) = (0_usize, nums.len() - 1);
        let mut sum = 0_i64;

        while lo < hi {
            sum += Self::gcd(nums[lo], nums[hi]) as i64;
            lo += 1;
            hi -= 1;
        }

        sum
    }
}
